### PR TITLE
feat: create quick search to jump to workspace or project

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[DET-7401]
User can jump to arbitrary workspace and project without going back to the list page.

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DET-7401]: https://determinedai.atlassian.net/browse/DET-7401?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        { signal: canceler.current.signal },
+      );
+      const filteredWorkspaces = workspaceResponse.workspaces.filter((w) => !w.immutable);
+      const projectAPIList = filteredWorkspaces
+        .map((workspace) =>
+          getWorkspaceProjects(
+            { id: workspace.id, sortBy: 'SORT_BY_NAME' },
+            { signal: canceler.current.signal },
+          ));
+      const projectResponse = (await Promise.all(projectAPIList))
```
Comment (keita-determined) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:42]: This is a concern. If user has lots of projects, we need to call the api a lot. Wanna make it 1 api call

Comment (trentwatt) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:42]: yeah ideally we could have one API call, we would need to make sure it respects RBAC though. 

@NicholasBlaskey if we wanted to make an API to get all projects in all workspaces for which a user is authorized, how might be go about doing that?

Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:42]: Thanks for the info. I'll make the change in different PR to make this PR simple

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  const onClickWorkspace = useCallback((workspaceId: number) => {
+    routeToReactUrl(paths.workspaceDetails(workspaceId));
+    onHideModal();
+  }, []);
+
+  const treeData: DefaultOptionType[] = useMemo(() => {
+    const data: DefaultOptionType[] = Array.from(workspaceMap)
+      .map(([ workspace, projects ]) => {
+        const treeChildren: DefaultOptionType[] = projects
+          .filter((project) => project.name.includes(searchText))
```
Comment (trentwatt) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx]: https://user-images.githubusercontent.com/28383080/187017218-8fcc27e7-d8b9-47e7-8471-95b235fe9d56.mov

i guess this is not ideal, but i think it's fine if we change the text of the input to say "Enter a project..."

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        )}
+        visible={isModalVisible}
+        width={'clamp(520px, 50vw, 1000px)'}
+        onCancel={onHideModal}>
+        <div className={css.modalBody}>
+          {isLoading ?
+            <Spinner center tip={'Loading...'} /> : (
+              <>
+                {treeData.length === 0 ?
+                  <Message title="No data found" type={MessageType.Empty} /> : (
```
Comment (trentwatt) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:143]: add some flavor? e.g. "No matching projects"

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    <div>
+      <div onClick={onShowModal}>
+        {children}
+      </div>
+      <Modal
+        closable={false}
+        footer={null}
+        title={(
+          <Input
+            placeholder="Search and Jump to workspace or project"
```
Comment (trentwatt) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:129]: "Search projects..." or similar

also if you pass the input a ref, you can call `focus()` on it with a use effect to put the cursor there by default.

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        silent: false,
+        type: ErrorType.Api,
+      });
+    } finally {
+      setIsLoading(false);
+    }
+  }, [ ]);
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setSearchText(e.target.value.toLocaleLowerCase());
```
Comment (trentwatt) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:66]: this changes as user is typing, no? 

probably can just keep the lowercasing to where you're doing the comparison

**Review by trentwatt — APPROVED** [review]
had some thoughts but LGTM

wuld be cool to highlight first matching entry and allow jump with enter

also small thing: search button remains highlighted after closing modal.

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        .map((workspace) =>
+          getWorkspaceProjects(
+            { id: workspace.id, sortBy: 'SORT_BY_NAME' },
+            { signal: canceler.current.signal },
+          ));
+      const projectResponse = (await Promise.all(projectAPIList))
+        .map((project) => project.projects);
+
+      // Promise.all preserves the order
+      const tempWorkspaceMap: Map<Workspace, Project[]> = new Map();
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:46]: yeah [here's](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps) the map vs object comparison

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+  const treeData: DefaultOptionType[] = useMemo(() => {
+    const data: DefaultOptionType[] = Array.from(workspaceMap)
+      .map(([ workspace, projects ]) => {
+        const treeChildren: DefaultOptionType[] = projects
+          .filter((project) => project.name.includes(searchText))
+          .map((project) => ({
+            title: (
+              <div className={`${css.flexRow} ${css.ellipsis}`}>
+                <Icon name="experiment" />
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:97]: I agree, but i think we dont have icon for project.

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx (hunk 2 above)
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx]: that wasn't intended. but fixed the bug. 
realized the bug after assign you as a reviewer 😅

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx (hunk 3 above)
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:146]: Definitely! I'll add `No matching workspace or projects` cuz i fixed the bug

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx (hunk 4 above)
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:132]: yes added

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx (hunk 5 above)
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:66]: it doesnt cuz i didnt add value property but its kinda vague so i'll change it

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+  const onClickWorkspace = useCallback((workspaceId: number) => {
+    routeToReactUrl(paths.workspaceDetails(workspaceId));
+    onHideModal();
+  }, []);
+
+  const treeData: DefaultOptionType[] = useMemo(() => {
+    const data: DefaultOptionType[] = Array.from(workspaceMap)
+      .map(([ workspace, projects ]) => {
+        const treeChildren: DefaultOptionType[] = projects
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:92]: i'll change it

File: webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        { limit: 0, sortBy: 'SORT_BY_NAME' },
+        { signal: canceler.current.signal },
+      );
+      const filteredWorkspaces = workspaceResponse.workspaces.filter((w) => !w.immutable);
+      const projectAPIList = filteredWorkspaces
+        .map((workspace) =>
+          getWorkspaceProjects(
+            { id: workspace.id, sortBy: 'SORT_BY_NAME' },
+            { signal: canceler.current.signal },
+          ));
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/pages/WorkspaceDetails/WorkspaceQuickSearch.tsx:41]: im not exactly sure what you mean here but i think it wouldn't work cuz getWorkspaceProjects returns promise functions here. but i agree that reverse + pop is not cool. I originally wanted to call popLeft with deque but js doesnt support it. so i can use `shift` which is the same time complexity O(n). Its not ideal to use linear time function but should be fine

**Comment by keita-determined** [comment]
Netlify CI keep running even though the log says its finished. All test cases passed, so i'll merge it

**Pull request merged**